// tdd-guard/dist/linters/rubocop/RuboCop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuboCop = void 0;
const runRuboCop_1 = require("./runRuboCop");
class RuboCop {
    run;
    constructor(run = runRuboCop_1.runRuboCop) {
        this.run = run;
    }
    async lint(filePaths, configPath) {
        const timestamp = new Date().toISOString();
        const args = buildArgs(filePaths, configPath);
        try {
            const { stdout } = await this.run(args, {
                shell: process.platform === 'win32',
            });
            return createLintData(timestamp, filePaths, parseResults(String(stdout)));
        }
        catch (error) {
            if (!isExecError(error))
                throw error;
            return createLintData(timestamp, filePaths, parseResults(String(error.stdout)));
        }
    }
}
exports.RuboCop = RuboCop;
// Helper functions
const buildArgs = (files, configPath) => {
    const args = [...files, '--format', 'json'];
    if (configPath) {
        args.push('--config', configPath);
    }
    return args;
};
const isExecError = (error) => error !== null && typeof error === 'object' && 'stdout' in error;
const parseResults = (stdout) => {
    try {
        const parsed = JSON.parse(stdout ?? '{"files":[]}');
        return parsed.files ?? [];
    }
    catch {
        return [];
    }
};
const createLintData = (timestamp, files, results) => {
    const issues = extractIssues(results);
    return {
        timestamp,
        files,
        issues,
        errorCount: countBySeverity(issues, 'error'),
        warningCount: countBySeverity(issues, 'warning'),
    };
};
const extractIssues = (results) => results.flatMap((file) => file.offenses.map(toIssue(file.path)));
const toIssue = (filePath) => (offense) => ({
    file: filePath,
    line: offense.location.line,
    column: offense.location.column,
    severity: mapSeverity(offense.severity),
    message: offense.message,
    rule: offense.cop_name,
});
const mapSeverity = (severity) => severity === 'error' || severity === 'fatal' ? 'error' : 'warning';
const countBySeverity = (issues, severity) => issues.filter((i) => i.severity === severity).length;
