// tdd-guard/dist/validation/models/AnthropicApi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnthropicApi = void 0;
const sdk_1 = __importDefault(require("@anthropic-ai/sdk"));
const Config_1 = require("../../config/Config");
const system_prompt_1 = require("../prompts/system-prompt");
class AnthropicApi {
    config;
    client;
    constructor(config) {
        this.config = config ?? new Config_1.Config();
        this.client = new sdk_1.default({
            apiKey: this.config.anthropicApiKey,
        });
    }
    async ask(prompt) {
        const response = await this.client.messages.create({
            model: this.config.modelVersion,
            system: system_prompt_1.SYSTEM_PROMPT,
            max_tokens: 1024,
            messages: [
                {
                    role: 'user',
                    content: prompt,
                },
            ],
        });
        return extractTextFromResponse(response);
    }
}
exports.AnthropicApi = AnthropicApi;
function extractTextFromResponse(response) {
    if (response.content.length === 0) {
        throw new Error('No content in response');
    }
    const firstContent = response.content[0];
    if (!('text' in firstContent) || !firstContent.text) {
        throw new Error('Response content does not contain text');
    }
    return firstContent.text;
}
