// tdd-guard/dist/validation/models/ClaudeCli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaudeCli = void 0;
const child_process_1 = require("child_process");
const path_1 = require("path");
const os_1 = require("os");
const fs_1 = require("fs");
const Config_1 = require("../../config/Config");
class ClaudeCli {
    config;
    constructor(config) {
        this.config = config ?? new Config_1.Config();
    }
    async ask(prompt) {
        const claudeBinary = this.getClaudeBinary();
        const args = [
            '-',
            '--output-format',
            'json',
            '--max-turns',
            '5',
            '--model',
            'sonnet',
            '--disallowed-tools',
            'TodoWrite',
            '--strict-mcp-config',
        ];
        const claudeDir = (0, path_1.join)(process.cwd(), '.claude');
        if (!(0, fs_1.existsSync)(claudeDir)) {
            (0, fs_1.mkdirSync)(claudeDir, { recursive: true });
        }
        const output = (0, child_process_1.execFileSync)(claudeBinary, args, {
            encoding: 'utf-8',
            timeout: 60000,
            input: prompt,
            cwd: claudeDir,
            shell: process.platform === 'win32',
        });
        // Parse the Claude CLI response and extract the result field
        const response = JSON.parse(output);
        return response.result;
    }
    getClaudeBinary() {
        if (this.config.useSystemClaude) {
            return 'claude';
        }
        return (0, path_1.join)((0, os_1.homedir)(), '.claude', 'local', 'claude');
    }
}
exports.ClaudeCli = ClaudeCli;
