// tdd-guard/dist/validation/models/ClaudeAgentSdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClaudeAgentSdk = void 0;
const Config_1 = require("../../config/Config");
const claude_agent_sdk_1 = require("@anthropic-ai/claude-agent-sdk");
const system_prompt_1 = require("../prompts/system-prompt");
class ClaudeAgentSdk {
    config;
    queryFn;
    constructor(config = new Config_1.Config(), queryFn = claude_agent_sdk_1.query) {
        this.config = config;
        this.queryFn = queryFn;
    }
    async ask(prompt) {
        const queryResult = this.queryFn({
            prompt,
            options: this.getQueryOptions(),
        });
        for await (const message of queryResult) {
            if (message.type !== 'result')
                continue;
            if (message.subtype === 'success') {
                if (message.is_error) {
                    throw new Error(message.result);
                }
                return message.result;
            }
            throw new Error(message.errors.length
                ? message.errors.join('; ')
                : `Claude Agent SDK error: ${message.subtype}`);
        }
        throw new Error('Claude Agent SDK error: No result message received');
    }
    getQueryOptions() {
        return {
            maxTurns: 1,
            systemPrompt: system_prompt_1.SYSTEM_PROMPT,
            allowedTools: [],
            disallowedTools: [
                'Read',
                'Edit',
                'MultiEdit',
                'Write',
                'Grep',
                'Glob',
                'Bash',
                'WebFetch',
                'WebSearch',
                'Task',
                'TodoWrite',
            ],
            thinking: { type: 'disabled' },
            model: this.config.modelVersion,
            permissionMode: 'dontAsk',
            strictMcpConfig: true,
            settingSources: [],
            persistSession: false,
        };
    }
}
exports.ClaudeAgentSdk = ClaudeAgentSdk;
