// @ast-grep/setup-lang/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.postinstall = postinstall;
exports.resolvePrebuild = resolvePrebuild;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
/**
 * Log to console
 */
function log(...args) {
    console.debug('@ast-grep/lang:', ...args);
}
/**
 * Move prebuild or build parser
 */
function postinstall(config) {
    const dir = config.dirname;
    const prebuild = resolvePrebuild(dir, true);
    if (prebuild) {
        log('prebuild found, do not need to build');
        return;
    }
    try {
        buildSrc(config);
    }
    catch (e) {
        log('build failed, please ensure tree-sitter-cli is installed as peer dependency');
        log(e);
    }
}
function buildSrc(config) {
    const { dirname } = config;
    const existing = node_path_1.default.join(dirname, 'src');
    if (!node_fs_1.default.existsSync(existing)) {
        log('tree-sitter src not found. If you are making a lang package, please run `pnpm source`.');
        return;
    }
    log('building parser from source');
    (0, node_child_process_1.execSync)('npm run build');
}
const PLATFORM_MAP = {
    darwin: 'macOS',
    linux: 'Linux',
    win32: 'Windows',
};
const ARCH_MAP = {
    x64: 'X64',
    arm64: 'ARM64',
};
/**
 * Resolve prebuild path
 */
function resolvePrebuild(dir, needLog = false) {
    const os = PLATFORM_MAP[process.platform];
    const arch = ARCH_MAP[process.arch];
    const prebuild = node_path_1.default.join(dir, 'prebuilds', `prebuild-${os}-${arch}`, 'parser.so');
    if (!os || !arch || !node_fs_1.default.existsSync(prebuild)) {
        if (needLog) {
            log(`no prebuild for ${os} ${arch}`);
        }
        return undefined;
    }
    if (needLog) {
        log(`found prebuild for ${os} ${arch}`);
    }
    return prebuild;
}
