// tdd-guard/dist/hooks/testCounter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.countTestDefinitions = countTestDefinitions;
const napi_1 = require("@ast-grep/napi");
const lang_python_1 = __importDefault(require("@ast-grep/lang-python"));
const lang_go_1 = __importDefault(require("@ast-grep/lang-go"));
const lang_ruby_1 = __importDefault(require("@ast-grep/lang-ruby"));
const lang_php_1 = __importDefault(require("@ast-grep/lang-php"));
const lang_rust_1 = __importDefault(require("@ast-grep/lang-rust"));
(0, napi_1.registerDynamicLanguage)({
    python: lang_python_1.default,
    go: lang_go_1.default,
    ruby: lang_ruby_1.default,
    php: lang_php_1.default,
    rust: lang_rust_1.default,
});
const JS_TEST_PATTERNS = [
    { pattern: 'test($$$A)' },
    { pattern: 'it($$$A)' },
    { pattern: 'test.skip($$$A)' },
    { pattern: 'test.only($$$A)' },
    { pattern: 'test.todo($$$A)' },
    { pattern: 'test.each($$$I)($$$A)' },
    { pattern: 'test.concurrent($$$A)' },
    { pattern: 'it.skip($$$A)' },
    { pattern: 'it.only($$$A)' },
    { pattern: 'it.todo($$$A)' },
    { pattern: 'it.each($$$I)($$$A)' },
    { pattern: 'it.concurrent($$$A)' },
];
function countJavaScriptTests(code, lang) {
    const ast = (0, napi_1.parse)(lang, code);
    const root = ast.root();
    const matches = new Set();
    for (const pattern of JS_TEST_PATTERNS) {
        for (const match of root.findAll({ rule: pattern })) {
            matches.add(match.range().start.line);
        }
    }
    return matches.size;
}
function countPythonTests(code) {
    const ast = (0, napi_1.parse)('python', code);
    const matches = ast.root().findAll({
        rule: {
            kind: 'function_definition',
            has: {
                field: 'name',
                regex: '^test_',
            },
        },
    });
    return matches.length;
}
function countGoTests(code) {
    const ast = (0, napi_1.parse)('go', code);
    const matches = ast.root().findAll({
        rule: {
            kind: 'function_declaration',
            has: {
                field: 'name',
                regex: '^(Test|Benchmark)',
            },
        },
    });
    return matches.length;
}
function countPhpTests(code) {
    const wrappedCode = wrapPhpSnippet(code);
    const ast = (0, napi_1.parse)('php', wrappedCode);
    const root = ast.root();
    const testLines = new Set();
    const byNaming = root.findAll({
        rule: {
            kind: 'method_declaration',
            has: {
                field: 'name',
                regex: '^test',
            },
        },
    });
    for (const match of byNaming) {
        testLines.add(match.range().start.line);
    }
    const byAttribute = root.findAll({
        rule: {
            kind: 'method_declaration',
            has: {
                kind: 'attribute_list',
                stopBy: 'end',
                has: {
                    kind: 'name',
                    regex: '^Test$',
                    stopBy: 'end',
                },
            },
        },
    });
    for (const match of byAttribute) {
        testLines.add(match.range().start.line);
    }
    return testLines.size;
}
function wrapPhpSnippet(code) {
    const needsTag = !code.trimStart().startsWith('<?php');
    const needsClass = !code.includes('class ');
    if (needsTag && needsClass) {
        return `<?php\nclass Wrapper {\n${code}\n}`;
    }
    if (needsTag) {
        return `<?php\n${code}`;
    }
    return code;
}
function countRubyTests(code) {
    const ast = (0, napi_1.parse)('ruby', code);
    const matches = ast.root().findAll({
        rule: {
            kind: 'call',
            has: {
                field: 'method',
                regex: '^it$',
            },
        },
    });
    return matches.length;
}
function countRustTests(code) {
    const ast = (0, napi_1.parse)('rust', code);
    const attributes = ast.root().findAll({
        rule: {
            pattern: '#[test]',
        },
    });
    return attributes.length;
}
function countTestDefinitions(code, language) {
    switch (language) {
        case 'javascript':
        case 'typescript':
            return countJavaScriptTests(code, language === 'typescript' ? napi_1.Lang.TypeScript : napi_1.Lang.JavaScript);
        case 'python':
            return countPythonTests(code);
        case 'go':
            return countGoTests(code);
        case 'php':
            return countPhpTests(code);
        case 'ruby':
            return countRubyTests(code);
        case 'rust':
            return countRustTests(code);
    }
}
