// tdd-guard/dist/hooks/userPromptHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserPromptHandler = void 0;
const GuardManager_1 = require("../guard/GuardManager");
const validationResults_1 = require("../contracts/validationResults");
class UserPromptHandler {
    guardManager;
    GUARD_COMMANDS = {
        ON: 'tdd-guard on',
        OFF: 'tdd-guard off',
    };
    constructor(guardManager) {
        this.guardManager = guardManager ?? new GuardManager_1.GuardManager();
    }
    async processUserCommand(hookData) {
        const data = JSON.parse(hookData);
        // Only process UserPromptSubmit events
        if (data.hook_event_name !== 'UserPromptSubmit') {
            return undefined;
        }
        const command = data.prompt?.toLowerCase();
        switch (command) {
            case this.GUARD_COMMANDS.ON:
                await this.guardManager.enable();
                return (0, validationResults_1.stopSession)('TDD Guard enabled');
            case this.GUARD_COMMANDS.OFF:
                await this.guardManager.disable();
                return (0, validationResults_1.stopSession)('TDD Guard disabled');
            default:
                return undefined;
        }
    }
    async getDisabledResult() {
        const isEnabled = await this.guardManager.isEnabled();
        if (!isEnabled) {
            return validationResults_1.allow;
        }
        return undefined;
    }
}
exports.UserPromptHandler = UserPromptHandler;
