// tdd-guard-vitest/dist/VitestReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VitestReporter = void 0;
const tdd_guard_1 = require("tdd-guard");
const node_path_1 = require("node:path");
class VitestReporter {
    storage;
    collectedData = new Map();
    constructor(options = {}) {
        this.storage =
            options.storage ??
                new tdd_guard_1.FileStorage(new tdd_guard_1.Config({ projectRoot: options.projectRoot }));
    }
    onTestModuleCollected(testModule) {
        this.collectedData.set(testModule.moduleId, {
            module: testModule,
            tests: [],
        });
    }
    onTestCaseResult(testCase) {
        const moduleId = testCase.module.moduleId;
        if (!moduleId)
            return;
        this.collectedData.get(moduleId)?.tests.push(testCase);
    }
    async onTestRunEnd(_testModules, unhandledErrors, reason) {
        // _testModules contains only module metadata, we use collected data from callbacks
        const formattedModules = formatAllModuleResults(this.collectedData);
        const output = createTestRunOutput(formattedModules, unhandledErrors, reason);
        await this.storage.saveTest(JSON.stringify(output, null, 2));
    }
}
exports.VitestReporter = VitestReporter;
function createTestRunOutput(testModules, unhandledErrors, reason) {
    return {
        testModules,
        unhandledErrors: unhandledErrors ?? [],
        ...(reason && { reason }),
    };
}
function formatAllModuleResults(collectedData) {
    return Array.from(collectedData.values()).map((data) => ({
        moduleId: data.module.moduleId,
        tests: moduleFailedToLoad(data)
            ? createTestForFailedModule(data)
            : formatNormalTests(data),
    }));
}
function moduleFailedToLoad(data) {
    return data.module.errors().length > 0 && data.tests.length === 0;
}
function createTestForFailedModule(data) {
    return [
        {
            name: (0, node_path_1.basename)(data.module.moduleId),
            fullName: data.module.moduleId,
            state: 'failed',
            errors: data.module.errors().map(formatError),
        },
    ];
}
function formatNormalTests(data) {
    return data.tests.map((test) => {
        const result = test.result();
        return {
            name: test.name,
            fullName: test.fullName,
            state: result.state === 'pending' ? 'skipped' : result.state,
            errors: result.errors?.map(formatError),
        };
    });
}
function formatError(error) {
    return {
        message: error.message,
        stack: error.stack,
        expected: error.expected,
        actual: error.actual,
    };
}
